### PR TITLE
TypeScript & type definitions generation

### PR DESCRIPTION
**Pull request by ramonsnir**
I recommend reading the section I added to `README.md`.

In short, other than switching the project itself to TypeScript:
- To play nice with TypeScript projects there should be some mechanism to codegen type definitions that will be used during development and build time. If `moduleType` is set to `typescript`, instead of a JS file a TS file is generated.
- Since config generation is only available for a specific environment, `--types-only-output` generates a file that has _only_ the types without an actual config. This generated file can and should be checked into Git, and will be later used for (a) CI jobs that build the codebase without being connected to any external service (b) if the code that generates the config needs the config type, this fixes 99% of the "bootstrap problem".
- Finally, since this is common to all of our projects, I incorporated the dynamic configuration part (e.g. reading secrets from AWS Secrets Manager or Parameter Store) into `single-config`. For users who wish it, the dynamic config will be included as part of the generated `config.js` (or `config.ts`). For users who don't wish it (e.g. us at Stackbit, with our long-running containers), the dynamic config isn't included in the output file but _is_ included in the generated type definitions.

I included a few examples in the `README.md` file, and you can see a real-world example in the non-public PR to our container project that uses all of these new features, in the manner described in the `README.md` file.

I'm not hooked on any of the names, so feel free to suggest new ones if you have _better_ alternatives.

Note after the PR feedback from my session with Simon: the definition of allowed environment selectors was moved from the parameters (CLI params or function arguments) to be part of the config object itself. It is tentatively called `"_envs": [...]`.

Known open items:

- [x] If the structure or types differ between envs, the inferred types are wrong since they are based only on the current environment. I hope to do this today, but I have to do something else first so we'll see.
- [ ] ~I am considering changing the dynamic config function to be immutable, so that it returns the additional values and then we use lodash to separately merge the two objects. This might be better later on.~

### CODE REVIEW

File: src/type-generation.ts [hunk 1]
```diff
@@ -0,0 +1,233 @@
+type Type =
```
Comment (ramonsnir) [inline src/type-generation.ts:1]: I recommend not having a heart attack.

I can add tests later, but I'm mostly really happy that I managed to write this entire file in one sitting and it just _worked_.

**Comment by ramonsnir** [comment]
@smnh Please review now :smile: I'm open, even very open, to a different name for `_envs`.

File: src/index.ts [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+        : extendedConfig;
+
+    const header = `// This file was automatically generated at ${new Date().toISOString()}`;
+    let moduleDefinition = `${header}
+module.exports = ${JSON.stringify(configToBeWritten, null, 4)};
+`;
+
+    await ensureDirectoryExistence(outputFilePath);
+
+    if (moduleType === 'globals') {
```
Comment (smnh) [inline src/index.ts]: wdyt about setting `moduleDefinition` with `module.exports` in its own if ... else clause like with other `moduleType`s 
```js
if (moduleType === 'node') {

} else if (moduleType === 'globals') {

} else if (moduleType === 'typescript') {

}
```

File: src/utils.ts [hunk 3]
```diff
@@ -0,0 +1,26 @@
+import path from 'path';
+import { promises as fs } from 'fs';
+
+export function logErrorAndThrow(error: Error): void {
+    console.error('\x1b[31m' + error.message + '\x1b[0m');
+    throw error;
+}
```
Comment (smnh) [inline src/utils.ts:7]: Is there advantage in receiving an Error object rather a message string and then creating a new error within this function like it was before?

In all instances where this function is used, it is called using same pattern - creating a `new Error` with a message string:
```js
logErrorAndThrow(
  new Error(
    message
  )
)
```

File: src/type-generation.ts (hunk 1 above)
Comment (smnh) [reply src/type-generation.ts:1]: will review this file some other day :)

**Review by smnh — CHANGES_REQUESTED** [review]

**Review by smnh — APPROVED** [review]

File: src/utils.ts (hunk 3 above)
Comment (ramonsnir) [reply src/utils.ts:7]: The error constructor is magical and sets the correct stack trace. That's why you must never throw non-Error objects (e.g. strings or `{ objects: ... }`), and why you must always call `new Error('foo')` as close to the place the error happened as possible.

File: src/index.ts (hunk 2 above)
Comment (ramonsnir) [reply src/index.ts]: Done

Comment (ramonsnir) [reply src/index.ts]: Done

**Pull request merged**